src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -245,7 +251,7 @@ private void InitializeConstructorParameters(ConstructorInfo constructorInfo)
 
                         // One object property cannot map to multiple constructor
                         // parameters (ConvertName above can't return multiple strings).
-                        parameterCache.Add(jsonParameterInfo.NameAsString, jsonParameterInfo);
+                        parameterCache.Add(jsonPropertyInfo.NameAsString!, jsonParameterInfo);
```

Comment:
nit: we should assert that `jsonPropertyInfo.NameAsString` and `jsonParameterInfo.NameAsString` are equal (I can do this in a follow up)